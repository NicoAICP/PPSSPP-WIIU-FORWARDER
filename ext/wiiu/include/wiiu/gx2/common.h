#pragma once

#define GX2_SCAN_BUFFER_ALIGNMENT      0x1000
#define GX2_SHADER_ALIGNMENT           0x100
#define GX2_CONTEXT_STATE_ALIGNMENT    0x100
#define GX2_UNIFORM_BLOCK_ALIGNMENT    0x100
#define GX2_DISPLAY_LIST_ALIGNMENT     0x20
#define GX2_VERTEX_BUFFER_ALIGNMENT    0x40
#define GX2_INDEX_BUFFER_ALIGNMENT     0x20

#define GX2_ENABLE                     TRUE
#define GX2_DISABLE                    FALSE

#define GX2_TRUE                       TRUE
#define GX2_FALSE                      FALSE

#ifndef GX2_COMP_SEL
#define GX2_COMP_SEL(c0, c1, c2, c3) (((c0) << 24) | ((c1) << 16) | ((c2) << 8) | (c3))
#ifdef __cplusplus
#include <functional>
#endif
#include <stdio.h>

#define _x 0
#define _y 1
#define _z 2
#define _w 3
#define _r 0
#define _g 1
#define _b 2
#define _a 3
#define _0 4
#define _1 5
#endif


#define GX2_DISABLE_WRAPS
