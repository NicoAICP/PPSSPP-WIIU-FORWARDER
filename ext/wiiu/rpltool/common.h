#ifndef COMMON_H
#define COMMON_H

#include <ctype.h>
#include <stdint.h>
#include <stdio.h>

typedef int8_t s8;
typedef int16_t s16;
typedef int32_t s32;

typedef uint8_t u8;
typedef uint16_t u16;
typedef uint32_t u32;

#define GET_ENUM_NAME(prefix, name, val) \
   case prefix##name: \
      return #name;

#define EMIT_ENUM_TO_STR_FUNC(name, cases) \
   static inline const char *name##_to_str(name type) { \
      switch (type) { cases } \
      return "unknown"; \
   }

#define GET_FLAG_NAME(prefix, name, val) \
   case prefix##name: \
      pos += snprintf(buffer + pos, sizeof(buffer) - pos, "%s%s", #name, seperator); \
      flags &= ~(1u << i); \
      break;

#define EMIT_FLAGS_TO_STR_FUNC(name, cases) \
   static inline const char *name##_to_str(int flags, const char *seperator) { \
      int i; \
      static char buffer[32 * 32]; \
      int pos = 0; \
      for (i = 0; i < 32; i++) \
         if ((1u << i) & flags) \
            switch (1u << i) { cases } \
      if (flags) \
         snprintf(buffer + pos, sizeof(buffer) - pos, "0x%08X", flags); \
      else if (pos) \
         buffer[pos - strlen(seperator)] = '\0'; \
      else \
         return "NULL"; \
      return buffer; \
   }

#define MAKE_MAGIC(a, b, c, d) (((a) << 24) | ((b) << 16) | ((c) << 8) | ((d) << 0))

#define log_var(X) printf("%-35s: 0x%0*X (%u)\n", #X, sizeof(X) * 2, (u32)(X), (u32)(X))
#define log_svar(X) printf("%-35s: 0x%0*X (%i)\n", #X, sizeof(X) * 2, (u32)(X), (s32)(X))
#define log_array(X) debug_array(X, sizeof(X), #X)
#define log_enum(type, X) printf("%-35s: 0x%X (%s)\n", #X, (X), type##_to_str(X))
#define log_flags(type, X) printf("%-35s: 0x%08X (%s)\n", #X, (X), type##_to_str(X, " | "))
#define log_magic(X) \
   printf("%-35s: '%c''%c''%c''%c' (0x%08X)\n", #X, (u32)(X) >> 24, (u32)(X) >> 16, (u32)(X) >> 8, (u32)(X), (u32)(X))
#define log_carray(X) debug_char_array(X, sizeof(X), #X)
#define log_str(X) printf("%-35s: \"%s\"\n", #X, (char *)(X))
#define log_off_str(X, data) printf("%-35s: 0x%08X \"%s\"\n", #X, (X), (char *)data + (X))
#define log_off_strs(X, data) \
   do { \
      printf("%-35s: 0x%08X ", #X, (X)); \
      const char *ptr = (char *)data + (X); \
      while (*ptr) { \
         ptr += printf("\"%s\" ", ptr) - 2; \
      } \
      printf("\n"); \
   } while (0)

static inline void dump_array(uint32_t *src, int size);
static inline void debug_array(uint32_t *src, int size, const char *name) {
   printf("%-35s:", name);
   dump_array(src, size);
}
static inline void debug_char_array(uint8_t *src, int size, const char *name) {
   int i;
   printf("%-35s: ", name);
   //   for (i = 0; i < size; i++)
   //      printf("%-2c ", isalnum(src[i])?src[i]: '?');
   //   printf("\n%-35s: ", "");
   for (i = 0; i < size; i++)
      printf("%02X ", src[i]);
   printf("\n");
}

static inline void dump_array(uint32_t *src, int size) {
   int i;
   size = (size / 4) - 1;
   printf("\n{");
   for (i = 0; i < size; i++) {
      if (!(i & 3))
         printf("\n   ");

      printf("0x%08X, ", src[i]);
   }

   if (!(i & 3))
      printf("\n   ");

   printf("0x%08X\n}\n", src[i]);
}

static inline u32 align_up(u32 x, u32 align) { return (x + (align - 1)) & ~(align - 1); }
static inline u32 align_down(u32 x, u32 align) { return x & ~(align - 1); }

#endif // COMMON_H
