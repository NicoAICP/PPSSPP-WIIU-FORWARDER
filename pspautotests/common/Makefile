TARGET_LIB = libcommon.a
EXTRA_TARGETS = libcommon_kernel.a
OBJS = common.o vram.o sysmem-imports.o
KERNEL_OBJS = common_kernel.o vram.o sysmem-imports.o

INCDIR = 
CFLAGS = -O0 -G 0 -g -Wall
CXXFLAGS = $(CFLAGS) -fno-exceptions -fno-rtti
ASFLAGS = $(CFLAGS)

LIBDIR =
LDFLAGS = -G0

PSPSDK=$(shell psp-config --pspsdk-path)
include $(PSPSDK)/lib/build.mak

%.o: %.S
	$(AS) $(ASFLAGS) -c -o $@ $<

common_kernel.o: common.c
	$(CC) $(CFLAGS) -DCOMMON_KERNEL=1 -c -o $@ $<

libcommon_kernel.a: $(KERNEL_OBJS)
	$(AR) cru $@ $(KERNEL_OBJS)
	$(RANLIB) $@
