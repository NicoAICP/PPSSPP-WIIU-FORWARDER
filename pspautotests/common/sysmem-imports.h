extern int sceKernelSetCompiledSdkVersion(int ver);
extern int sceKernelSetCompiledSdkVersion370(int ver);
extern int sceKernelSetCompiledSdkVersion380_390(int ver);
extern int sceKernelSetCompiledSdkVersion395(int ver);
extern int sceKernelSetCompiledSdkVersion401_402(int ver);
extern int sceKernelSetCompiledSdkVersion500_505(int ver);
extern int sceKernelSetCompiledSdkVersion507(int ver);
extern int sceKernelSetCompiledSdkVersion600_602(int ver);
extern int sceKernelSetCompiledSdkVersion603_605(int ver);
extern int sceKernelSetCompiledSdkVersion606(int ver);
extern int sceKernelGetCompiledSdkVersion();