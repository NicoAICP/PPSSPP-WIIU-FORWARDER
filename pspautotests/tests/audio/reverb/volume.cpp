#include <common.h>
#include <pspkernel.h>
#include <psputility.h>
#include "../sascore/sascore.h"

static const int SAMPLE_COUNT = 16384;

__attribute__((aligned(64))) SasCore sasCore;
__attribute__((aligned(64))) short pcm[256 * 2 * 16] = { 0 };
__attribute__((aligned(64))) short samples[SAMPLE_COUNT * 2 * 16] = { 0 };

void printSample(int i) {
	schedf("  [%03x] %04x %04x\n", i, (u16)samples[i * 2 + 0], (u16)samples[i * 2 + 1]);
}

void testPCMOutput(const char *title, int dry, int wet, int type, short pcmVol, short effectVol, short revVol) {
	__sceSasRevVON(&sasCore, dry, wet);
	__sceSasSetVolume(&sasCore, 0, pcmVol, pcmVol, effectVol, effectVol);
	__sceSasRevEVOL(&sasCore, revVol, revVol);
	__sceSasRevParam(&sasCore, 0, 0);
	__sceSasRevType(&sasCore, type);
	__sceSasSetKeyOff(&sasCore, 0);
	__sceSasSetADSRmode(&sasCore, 0, 15, PSP_SAS_ADSR_CURVE_MODE_LINEAR_INCREASE, PSP_SAS_ADSR_CURVE_MODE_DIRECT, PSP_SAS_ADSR_CURVE_MODE_DIRECT, PSP_SAS_ADSR_CURVE_MODE_DIRECT);
	__sceSasSetADSR(&sasCore, 0, 15, 0x40000000, 0x40000000, 0x40000000, 0x40000000);

	const int grainSize = 256;
	__sceSasSetGrain(&sasCore, grainSize);
	__sceSasSetOutputmode(&sasCore, 0);
	__sceSasSetVoicePCM(&sasCore, 0, pcm, 256, 0);
	__sceSasSetKeyOn(&sasCore, 0);

	memset(samples, 0xcc, sizeof(samples));

	sceKernelDelayThread(50000);

	bool foundInteresting = false;
	int interestingTimeout = 50;
	while (!foundInteresting && --interestingTimeout > 0) {
		__sceSasCore(&sasCore, samples);
		for (int i = 0; i < grainSize; ++i) {
			if (samples[i] != 0 && samples[i] != 0x4000) {
				foundInteresting = true;
				break;
			}
		}
	}

	for (int off = grainSize; off < SAMPLE_COUNT; off += grainSize) {
		__sceSasCore(&sasCore, samples + off * 2);
	}

	double deltaSquared = 0.0;
	for (int i = 0; i < SAMPLE_COUNT; i += 2) {
		deltaSquared += (s16)samples[i * 2] * (s16)samples[i * 2];
	}

	double rms = sqrt(deltaSquared / (double)(SAMPLE_COUNT / 2));
	double rmsRatio = rms / 16384.0;
	if (rmsRatio >= 0.98 && rmsRatio <= 1.02) {
		checkpoint("%s: root mean is near 1x input", title);
	} else if (rmsRatio >= 1.96 && rmsRatio <= 2.04) {
		checkpoint("%s: root mean is near 2x input", title);
	} else {
		checkpoint("%s: root mean is not near 1x or 2x input", title);
	}
}

extern "C" int main(int argc, char *argv[]) {
	sceUtilityLoadModule(PSP_MODULE_AV_AVCODEC);
	sceUtilityLoadModule(PSP_MODULE_AV_SASCORE);

	__sceSasInit(&sasCore, 128, 32, 1, 44100);

	for (size_t i = 0; i < ARRAY_SIZE(pcm); ++i) {
		pcm[i] = 0;
	}

	for (size_t i = 0; i < 256; ++i) {
		pcm[i] = 16384;
	}

	checkpointNext("Deltas at full vol");
	testPCMOutput("  OFF", 1, 1, PSP_SAS_EFFECT_TYPE_OFF, 0x1000, 0x1000, 0x1000);
	testPCMOutput("  ROOM", 1, 1, PSP_SAS_EFFECT_TYPE_ROOM, 0x1000, 0x1000, 0x1000);
	testPCMOutput("  UNK1", 1, 1, PSP_SAS_EFFECT_TYPE_UNK1, 0x1000, 0x1000, 0x1000);
	testPCMOutput("  UNK2", 1, 1, PSP_SAS_EFFECT_TYPE_UNK2, 0x1000, 0x1000, 0x1000);
	testPCMOutput("  UNK3", 1, 1, PSP_SAS_EFFECT_TYPE_UNK3, 0x1000, 0x1000, 0x1000);
	testPCMOutput("  HALL", 1, 1, PSP_SAS_EFFECT_TYPE_HALL, 0x1000, 0x1000, 0x1000);
	testPCMOutput("  SPACE", 1, 1, PSP_SAS_EFFECT_TYPE_SPACE, 0x1000, 0x1000, 0x1000);
	testPCMOutput("  ECHO", 1, 1, PSP_SAS_EFFECT_TYPE_ECHO, 0x1000, 0x1000, 0x1000);
	testPCMOutput("  DELAY", 1, 1, PSP_SAS_EFFECT_TYPE_DELAY, 0x1000, 0x1000, 0x1000);
	testPCMOutput("  PIPE", 1, 1, PSP_SAS_EFFECT_TYPE_PIPE, 0x1000, 0x1000, 0x1000);

	checkpointNext("Deltas with dry=0");
	testPCMOutput("  OFF", 0, 1, PSP_SAS_EFFECT_TYPE_OFF, 0x1000, 0x1000, 0x1000);
	testPCMOutput("  ROOM", 0, 1, PSP_SAS_EFFECT_TYPE_ROOM, 0x1000, 0x1000, 0x1000);
	testPCMOutput("  UNK1", 0, 1, PSP_SAS_EFFECT_TYPE_UNK1, 0x1000, 0x1000, 0x1000);
	testPCMOutput("  UNK2", 0, 1, PSP_SAS_EFFECT_TYPE_UNK2, 0x1000, 0x1000, 0x1000);
	testPCMOutput("  UNK3", 0, 1, PSP_SAS_EFFECT_TYPE_UNK3, 0x1000, 0x1000, 0x1000);
	testPCMOutput("  HALL", 0, 1, PSP_SAS_EFFECT_TYPE_HALL, 0x1000, 0x1000, 0x1000);
	testPCMOutput("  SPACE", 0, 1, PSP_SAS_EFFECT_TYPE_SPACE, 0x1000, 0x1000, 0x1000);
	testPCMOutput("  ECHO", 0, 1, PSP_SAS_EFFECT_TYPE_ECHO, 0x1000, 0x1000, 0x1000);
	testPCMOutput("  DELAY", 0, 1, PSP_SAS_EFFECT_TYPE_DELAY, 0x1000, 0x1000, 0x1000);
	testPCMOutput("  PIPE", 0, 1, PSP_SAS_EFFECT_TYPE_PIPE, 0x1000, 0x1000, 0x1000);

	checkpointNext("Deltas with mainVol=0");
	testPCMOutput("  OFF", 1, 1, PSP_SAS_EFFECT_TYPE_OFF, 0x0000, 0x1000, 0x1000);
	testPCMOutput("  ROOM", 1, 1, PSP_SAS_EFFECT_TYPE_ROOM, 0x0000, 0x1000, 0x1000);
	testPCMOutput("  UNK1", 1, 1, PSP_SAS_EFFECT_TYPE_UNK1, 0x0000, 0x1000, 0x1000);
	testPCMOutput("  UNK2", 1, 1, PSP_SAS_EFFECT_TYPE_UNK2, 0x0000, 0x1000, 0x1000);
	testPCMOutput("  UNK3", 1, 1, PSP_SAS_EFFECT_TYPE_UNK3, 0x0000, 0x1000, 0x1000);
	testPCMOutput("  HALL", 1, 1, PSP_SAS_EFFECT_TYPE_HALL, 0x0000, 0x1000, 0x1000);
	testPCMOutput("  SPACE", 1, 1, PSP_SAS_EFFECT_TYPE_SPACE, 0x0000, 0x1000, 0x1000);
	testPCMOutput("  ECHO", 1, 1, PSP_SAS_EFFECT_TYPE_ECHO, 0x0000, 0x1000, 0x1000);
	testPCMOutput("  DELAY", 1, 1, PSP_SAS_EFFECT_TYPE_DELAY, 0x0000, 0x1000, 0x1000);
	testPCMOutput("  PIPE", 1, 1, PSP_SAS_EFFECT_TYPE_PIPE, 0x0000, 0x1000, 0x1000);

	checkpointNext("Deltas with wet=0");
	testPCMOutput("  OFF", 1, 0, PSP_SAS_EFFECT_TYPE_OFF, 0x1000, 0x1000, 0x1000);
	testPCMOutput("  ROOM", 1, 0, PSP_SAS_EFFECT_TYPE_ROOM, 0x1000, 0x1000, 0x1000);
	testPCMOutput("  UNK1", 1, 0, PSP_SAS_EFFECT_TYPE_UNK1, 0x1000, 0x1000, 0x1000);
	testPCMOutput("  UNK2", 1, 0, PSP_SAS_EFFECT_TYPE_UNK2, 0x1000, 0x1000, 0x1000);
	testPCMOutput("  UNK3", 1, 0, PSP_SAS_EFFECT_TYPE_UNK3, 0x1000, 0x1000, 0x1000);
	testPCMOutput("  HALL", 1, 0, PSP_SAS_EFFECT_TYPE_HALL, 0x1000, 0x1000, 0x1000);
	testPCMOutput("  SPACE", 1, 0, PSP_SAS_EFFECT_TYPE_SPACE, 0x1000, 0x1000, 0x1000);
	testPCMOutput("  ECHO", 1, 0, PSP_SAS_EFFECT_TYPE_ECHO, 0x1000, 0x1000, 0x1000);
	testPCMOutput("  DELAY", 1, 0, PSP_SAS_EFFECT_TYPE_DELAY, 0x1000, 0x1000, 0x1000);
	testPCMOutput("  PIPE", 1, 0, PSP_SAS_EFFECT_TYPE_PIPE, 0x1000, 0x1000, 0x1000);

	checkpointNext("Deltas with effectVol=0");
	testPCMOutput("  OFF", 1, 1, PSP_SAS_EFFECT_TYPE_OFF, 0x1000, 0x0000, 0x1000);
	testPCMOutput("  ROOM", 1, 1, PSP_SAS_EFFECT_TYPE_ROOM, 0x1000, 0x0000, 0x1000);
	testPCMOutput("  UNK1", 1, 1, PSP_SAS_EFFECT_TYPE_UNK1, 0x1000, 0x0000, 0x1000);
	testPCMOutput("  UNK2", 1, 1, PSP_SAS_EFFECT_TYPE_UNK2, 0x1000, 0x0000, 0x1000);
	testPCMOutput("  UNK3", 1, 1, PSP_SAS_EFFECT_TYPE_UNK3, 0x1000, 0x0000, 0x1000);
	testPCMOutput("  HALL", 1, 1, PSP_SAS_EFFECT_TYPE_HALL, 0x1000, 0x0000, 0x1000);
	testPCMOutput("  SPACE", 1, 1, PSP_SAS_EFFECT_TYPE_SPACE, 0x1000, 0x0000, 0x1000);
	testPCMOutput("  ECHO", 1, 1, PSP_SAS_EFFECT_TYPE_ECHO, 0x1000, 0x0000, 0x1000);
	testPCMOutput("  DELAY", 1, 1, PSP_SAS_EFFECT_TYPE_DELAY, 0x1000, 0x0000, 0x1000);
	testPCMOutput("  PIPE", 1, 1, PSP_SAS_EFFECT_TYPE_PIPE, 0x1000, 0x0000, 0x1000);

	checkpointNext("Deltas with reverbVol=0");
	testPCMOutput("  OFF", 1, 1, PSP_SAS_EFFECT_TYPE_OFF, 0x1000, 0x1000, 0x0000);
	testPCMOutput("  ROOM", 1, 1, PSP_SAS_EFFECT_TYPE_ROOM, 0x1000, 0x1000, 0x0000);
	testPCMOutput("  UNK1", 1, 1, PSP_SAS_EFFECT_TYPE_UNK1, 0x1000, 0x1000, 0x0000);
	testPCMOutput("  UNK2", 1, 1, PSP_SAS_EFFECT_TYPE_UNK2, 0x1000, 0x1000, 0x0000);
	testPCMOutput("  UNK3", 1, 1, PSP_SAS_EFFECT_TYPE_UNK3, 0x1000, 0x1000, 0x0000);
	testPCMOutput("  HALL", 1, 1, PSP_SAS_EFFECT_TYPE_HALL, 0x1000, 0x1000, 0x0000);
	testPCMOutput("  SPACE", 1, 1, PSP_SAS_EFFECT_TYPE_SPACE, 0x1000, 0x1000, 0x0000);
	testPCMOutput("  ECHO", 1, 1, PSP_SAS_EFFECT_TYPE_ECHO, 0x1000, 0x1000, 0x0000);
	testPCMOutput("  DELAY", 1, 1, PSP_SAS_EFFECT_TYPE_DELAY, 0x1000, 0x1000, 0x0000);
	testPCMOutput("  PIPE", 1, 1, PSP_SAS_EFFECT_TYPE_PIPE, 0x1000, 0x1000, 0x0000);

	return 0;
}