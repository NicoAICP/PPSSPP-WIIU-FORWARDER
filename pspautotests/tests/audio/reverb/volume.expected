[x] Deltas at full vol
[r]   OFF: root mean is near 2x input
[r]   ROOM: root mean is not near 1x or 2x input
[r]   UNK1: root mean is not near 1x or 2x input
[r]   UNK2: root mean is not near 1x or 2x input
[r]   UNK3: root mean is not near 1x or 2x input
[r]   HALL: root mean is not near 1x or 2x input
[r]   SPACE: root mean is near 1x input
[r]   ECHO: root mean is not near 1x or 2x input
[r]   DELAY: root mean is not near 1x or 2x input
[r]   PIPE: root mean is not near 1x or 2x input

[x] Deltas with dry=0
[r]   OFF: root mean is near 1x input
[r]   ROOM: root mean is not near 1x or 2x input
[r]   UNK1: root mean is not near 1x or 2x input
[r]   UNK2: root mean is not near 1x or 2x input
[r]   UNK3: root mean is not near 1x or 2x input
[r]   HALL: root mean is not near 1x or 2x input
[r]   SPACE: root mean is not near 1x or 2x input
[r]   ECHO: root mean is not near 1x or 2x input
[r]   DELAY: root mean is not near 1x or 2x input
[r]   PIPE: root mean is not near 1x or 2x input

[x] Deltas with mainVol=0
[r]   OFF: root mean is near 1x input
[r]   ROOM: root mean is not near 1x or 2x input
[r]   UNK1: root mean is not near 1x or 2x input
[r]   UNK2: root mean is not near 1x or 2x input
[r]   UNK3: root mean is not near 1x or 2x input
[r]   HALL: root mean is not near 1x or 2x input
[r]   SPACE: root mean is not near 1x or 2x input
[r]   ECHO: root mean is not near 1x or 2x input
[r]   DELAY: root mean is not near 1x or 2x input
[r]   PIPE: root mean is not near 1x or 2x input

[x] Deltas with wet=0
[r]   OFF: root mean is near 1x input
[r]   ROOM: root mean is near 1x input
[r]   UNK1: root mean is near 1x input
[r]   UNK2: root mean is near 1x input
[r]   UNK3: root mean is near 1x input
[r]   HALL: root mean is near 1x input
[r]   SPACE: root mean is near 1x input
[r]   ECHO: root mean is near 1x input
[r]   DELAY: root mean is near 1x input
[r]   PIPE: root mean is near 1x input

[x] Deltas with effectVol=0
[r]   OFF: root mean is near 1x input
[r]   ROOM: root mean is near 1x input
[r]   UNK1: root mean is near 1x input
[r]   UNK2: root mean is near 1x input
[r]   UNK3: root mean is near 1x input
[r]   HALL: root mean is near 1x input
[r]   SPACE: root mean is near 1x input
[r]   ECHO: root mean is near 1x input
[r]   DELAY: root mean is near 1x input
[r]   PIPE: root mean is near 1x input

[x] Deltas with reverbVol=0
[r]   OFF: root mean is near 1x input
[r]   ROOM: root mean is near 1x input
[r]   UNK1: root mean is near 1x input
[r]   UNK2: root mean is near 1x input
[r]   UNK3: root mean is near 1x input
[r]   HALL: root mean is near 1x input
[r]   SPACE: root mean is near 1x input
[r]   ECHO: root mean is near 1x input
[r]   DELAY: root mean is near 1x input
[r]   PIPE: root mean is near 1x input
